File: android/common/compose/src/main/java/com/gchristov/newsfeed/commoncompose/theme/Theme.kt
package com.gchristov.newsfeed.commoncompose.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider

@Composable
fun Theme(
    type: ThemeType = if (isSystemInDarkTheme()) ThemeType.Dark else ThemeType.Light,
    content: @Composable () -> Unit
) {
    CompositionLocalProvider(
        LocalBackgrounds provides type.backgrounds(),
        LocalContentColors provides type.contentColors(),
        LocalTypography provides typography(),
        LocalShapes provides shapes(),
    ) {
        MaterialTheme(content = content)
    }
}

object Theme {
    val backgrounds: Backgrounds
        @Composable
        get() = LocalBackgrounds.current

    val contentColors: ContentColors
        @Composable
        get() = LocalContentColors.current

    val typography: Typography
        @Composable
        get() = LocalTypography.current

    val shapes: Shapes
        @Composable
        get() = LocalShapes.current
}

sealed class ThemeType {
    object Light : ThemeType()
    object Dark : ThemeType()

    @Composable
    fun backgrounds() = when (this) {
        Dark -> darkBackgrounds()
        Light -> lightBackgrounds()
    }

    @Composable
    fun contentColors() = when (this) {
        Dark -> darkContentColors()
        Light -> lightContentColors()
    }
}

File: ios/common-swiftui/Sources/Theme/Theme.swift
import Foundation

public class Theme: ObservableObject {
    @Published public var type: ThemeType
    public lazy var backgrounds: Backgrounds = type.backgrounds()
    public lazy var contentColors: ContentColors = type.contentColors()
    public lazy var typography: Typography = type.typography()
    public lazy var shapes: Shapes = type.shapes()
    
    public init(type: ThemeType) {
        self.type = type
    }
}

public enum ThemeType {
    case dark
    case light
    
    func backgrounds() -> Backgrounds {
        switch self {
        case .dark: return darkBackgrounds()
        case .light: return lightBackgrounds()
        }
    }
    
    func contentColors() -> ContentColors {
        switch self {
        case .dark: return darkContentColors()
        case .light: return lightContentColors()
        }
    }
    
    func typography() -> Typography {
        return CommonSwiftUi.typography()
    }
    
    func shapes() -> Shapes {
        return CommonSwiftUi.shapes()
    }
}
